### ISSUE TITLE
'docker ps -s' hangs on CentOS 7.2 with Docker 1.11

### ISSUE THREAD

**Issue by jerbia** [docker/docker#22116]
---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------
Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST
-->

**Output of `docker version`:**

```
# docker version
Client:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:40:36 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.0
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   4dc5990
 Built:        Wed Apr 13 18:40:36 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
# docker info
 Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 2
Server Version: 1.11.0
Storage Driver: devicemapper
 Pool Name: docker-8:1-33690131-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 473.6 MB
 Data Space Total: 107.4 GB
 Data Space Available: 7.82 GB
 Metadata Space Used: 1.053 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.146 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2015-12-01)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 3.10.0-327.13.1.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.305 GiB
Name: azy-centos-72
ID: HKFD:YLCX:MZPD:GVKJ:2R72:RF47:3DDP:ECVU:MTS2:JMVL:OQ5S:YBHN
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Username: xxxxxxxxx
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
 
# uname -a
Linux azy-centos-72 3.10.0-327.13.1.el7.x86_64 #1 SMP Thu Mar 31 16:04:38 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

# cat /etc/centos-release
CentOS Linux release 7.2.1511 (Core)

**Steps to reproduce the issue:**
1. Install Docker 1.11
2. Run a container: 'docker run --rm -it busybox /bin/sh' 
3. Run 'docker ps -s' --> takes 40 seconds to return
4. In the meanwhile, open another terminal, run 'docker ps' --> takes ~30 seconds to respond

Note that this does not happen to me on Ubuntu 14.04 (devicemapper issue?).

**Describe the results you received:**
'docker ps -s' hangs for 40 seconds. Any other command that you try to run at the same time - also hangs for 40 seconds.

**Describe the results you expected:**
'docker ps -s' should return immediately...

**Additional information you deem important (e.g. issue happens only occasionally):**
Happens every time, on different CentOS 7.2 machines.

**Comment by cpuguy83** [comment]
@jerbia `-s` has to calculate the size of the rootfs which can indeed take some time.
The container is locked during this time which means any other command that needs to access the container (like a subsequent `docker ps`) will be blocked on that container until it is unlocked.

With devicemapper it also has to figure out what the diff is before it can calculate the size of the diff.
AUFS doesn't need to figure out what the diff is since it's always available.

**Comment by mlaventure** [comment]
@cpuguy83 @jerbia 

I can reproduce it. And I get the following in the logs:
```
ERRO[2455] devmapper: Error unmounting device 20ff369e9f0138f634088f5e4e85d367e5620c35f1bb8b485d81c17757a614a2: Device is Busy
```

We basically timeout when executing one of the `defer driver.Put(id)` statements @ https://github.com/docker/docker/blob/v1.11.0/daemon/graphdriver/fsdiff.go#L97

Not being so familiar with devmapper, I can't tell why it would be busy for so long (the timeout is 20s).

**Comment by cpuguy83** [comment]
Ok, I think this might because because fsdiff is calling Get/Put willy-nilly  and there's no real protection in the driver to keep `Put` from unmounting.

**Comment by cpuguy83** [comment]
I can make a patch for 1.11.1 that adds a reccounting for Get/Put so we don't hit this situation.
For 1.12 we'll need a better solution though that'll likely too big a change for 1.11.1.

**Comment by thaJeztah** [comment]
@cpuguy83 if you have a fix for 1.11, let me assign you :smile:

**Comment by daniel-smith-kaazing** [comment]
We are seeing the same on CentOS with docker 1.12.1. Docker becomes unusable even for simple tasks - remove container, remove images, builds, etc.
Here's the info:
docker --version
Docker version 1.12.1, build 23cf638
 uname -a
Linux YOUR_HOSTNAME_HERE 3.10.0-327.36.1.el7.x86_64 #1 SMP Sun Sep 18 13:04:29 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

**Comment by mlaventure** [comment]
@daniel-smith-kaazing please open a new issue with as much information as you can (e.g. daemon logs, way to reproduce, ...).
